File: python/sqrt.py
# Time -> 0(log(N)) Space 0(1)
def mySqrt(x: int) -> int:
    start, end = 0, x
    best = 0
    while start <= end:
        mid = (start + end) // 2
        midSquare = mid * mid
        if midSquare == x:
            return mid
        elif midSquare > x:
            end = mid - 1
        else:
            start = mid + 1

        if midSquare < x:
            best = mid

    return best


print(mySqrt(8))

File: go/Sqrt/sqrt.go
package main

import "fmt"

// Time -> 0(log(N)) Space -> 0(1)
func mySqrt(x int) int {
	start, stop := 0, x
	best := 0
	for start <= stop {
		mid := (start + stop) / 2
		midSquare := mid * mid
		if midSquare == x {
			return mid
		} else if midSquare > x {
			stop = mid - 1
		} else {
			start = mid + 1
		}
		if midSquare < x {
			best = mid
		}
	}
	return best
}

func main() {
	fmt.Println(mySqrt(15))
}
